src/compiler/checker.ts

```diff
@@ -28577,13 +28577,17 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
 
     function getTypeOfPropertyOfContextualType(type: Type, name: __String, nameType?: Type) {
         return mapType(type, t => {
-            if (isGenericMappedType(t) && !t.declaration.nameType) {
-                const constraint = getConstraintTypeFromMappedType(t);
-                const constraintOfConstraint = getBaseConstraintOfType(constraint) || constraint;
-                const propertyNameType = nameType || getStringLiteralType(unescapeLeadingUnderscores(name));
-                if (isTypeAssignableTo(propertyNameType, constraintOfConstraint)) {
-                    return substituteIndexedMappedType(t, propertyNameType);
-                }
+            if (everyContainedType(t, t => isGenericMappedType(t) && !t.declaration.nameType)) {
+                const newTypes = mapDefined(t.flags & TypeFlags.Intersection ? (t as IntersectionType).types : [t], t => {
+                    const mappedType = t as MappedType;
+                    const constraint = getConstraintTypeFromMappedType(mappedType);
+                    const constraintOfConstraint = getBaseConstraintOfType(constraint) || constraint;
+                    const propertyNameType = nameType || getStringLiteralType(unescapeLeadingUnderscores(name));
+                    if (isTypeAssignableTo(propertyNameType, constraintOfConstraint)) {
+                        return substituteIndexedMappedType(mappedType, propertyNameType);
+                    }
+                });
+                return newTypes.length ? getIntersectionType(newTypes) : undefined;

```

Comment:
This will quietly elide intersection members whose substitution fails the `isTypeAssignableTo(propertyNameType, constraintOfConstraint)` check. I honestly can't think of how that could occur offhand, but I think to better mimic old behavior, since intersections _are_ structured types, and would previously only take the other branch. If `newTypes.length !== (t.flags & TypeFlags.Intersection ? (t as IntersectionType).types : [t]).length`, we need to fall through to the `StructuredType` fallback below.

Reply:
I've tried a couple of things to hit such a case and I can't think of any. I can introduce the proposed change but personally I think that it should be accompanied with a test case. I understand the desire to match the old behavior but:
1. if we manage to grab a concrete property in the branch responsible for structured types then this is basically the same thing as substituting indexed mapped type (so doing it within this branch responsible for generic mapped types leads to the same results already)
2. if we manage to grab some index info there then I find it extremely unlikely there is a way to reach that branch of code if all intersected mapped type failed `isTypeAssignableTo(propertyNameType, constraintOfConstraint)` check. In a lot of situations, the `constraitnOfConstraint` is already `string | number | symbol` and even when it's not... the index type has to come from one of the intersected members, so some of them had to pass this `isTypeAssignableTo` check and thus `newTypes` wouldn't be empty.
3. The `isTupleType` is just impossible to be hit by intersections

So, all in all, I think that the requested change wouldn't be meaningful - but, ofc, I could be wrong here. And even if such intersections could today return something meaningful then that, I think, would be inconsistent with the `isGenericMappedType(t) && !t.declaration.nameType` branch. That's because that branch doesn't fallback to the branch related to structured types - even if `isTypeAssignableTo(propertyNameType, constraintOfConstraint)`. So if this doesn't happen for generic mapped types - why should it happen for intersected ones? That would be an inconsistent behavior to me.